pkg/controller/elasticsearch/driver/pvc.go

```diff
@@ -0,0 +1,122 @@
+package driver
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	"github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1alpha1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/label"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/sset"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/stringsutil"
+)
+
+// GarbageCollectPVCs ensures PersistentVolumeClaims created for the given es resource are deleted
+// when no longer used, since this is not done automatically by the StatefulSet controller.
+// Related issue in the k8s repo: https://github.com/kubernetes/kubernetes/issues/55045
+// It sets an owner reference to automatically delete PVCs on es deletion, and garbage collects
+// unused PVCs of existing StatefulSets.
+// Note we do **not** delete the corresponding PersistentVolumes but just the PersistentVolumeClaims.
+// PV deletion is left to the responsibility of the storage class reclaim policy.
+func GarbageCollectPVCs(
+	k8sClient k8s.Client,
+	scheme *runtime.Scheme,
+	es v1alpha1.Elasticsearch,
+	actualStatefulSets sset.StatefulSetList,
+	expectedStatefulSets sset.StatefulSetList,
+) error {
+	// PVCs are using the same labels as their corresponding StatefulSet, so we can filter on ES cluster name.
+	var pvcs corev1.PersistentVolumeClaimList
+	if err := k8sClient.List(&client.ListOptions{
+		Namespace:     es.Namespace,
+		LabelSelector: label.NewLabelSelectorForElasticsearch(es),
+	}, &pvcs); err != nil {
+		return err
+	}
+	if err := reconcilePVCsOwnerRef(k8sClient, scheme, es, pvcs.Items); err != nil {
+		return err
+	}
+	return deleteUnusedPVCs(k8sClient, pvcs.Items, actualStatefulSets, expectedStatefulSets)
+}
+
+// reconcilePVCsOwnerRef ensures PVCs created for this Elasticsearch cluster have an owner ref set to
+// the Elasticsearch resource, so they are deleted automatically upon Elasticsearch deletion.
+// A subtle race condition exists: users may still end up with leftover PVCs if the Elasticsearch resource
+// gets deleted right after creation or update, before this is called.
```

Comment:
To handle this we could run a gc function outside the scope of a single cluster reconciliation (ie. working with all clusters), removing resources not related to any existing cluster.

I think the chance that race condition occurs is rather low (update the ES resource so we get new PVCs created, then immediately stop or disconnect the operator before it gets a chance to patch newly created PVCs). So may not be worth optimising for. And even if it may happen, then it's probably not a big deal. Thoughts?

Reply:
I also think it's unlikely but I'm constantly reproducing the race with: `k apply -f es.yml; sleep 1; k delete es --all`. By reapplying the ES resource, the PV are reused.

Reply:
@anyasabo [suggestion](https://github.com/elastic/cloud-on-k8s/issues/1288) to set the PVC ownerRef within the claim template in the StatefulSet spec should allow us to work around that. No need to reconcile owner ref after the fact, PVCs get created using owner ref specified in the StatefulSet spec at creation time.